src/OrleansEventSourcing/EventStorage/LogViewAdaptor.cs

```diff
@@ -0,0 +1,373 @@
+namespace Orleans.EventSourcing.EventStorage
+{
+
+    /// <summary>
+    /// A log view adaptor that is based on an event store
+    /// </summary>
+    /// <typeparam name="TState">The type of the log view, or grain state.</typeparam>
+    /// <typeparam name="TEvent">The type of the events (usually just object)</typeparam>
+    internal class EventStoreLogViewAdaptor<TState, TEvent> : PrimaryBasedLogViewAdaptor<TState, TEvent, SubmissionEntryWithGuid<TEvent>>
+        where TState : class, new() where TEvent : class
+    {
+        // the provider that created this adaptor
+        private LogConsistencyProvider provider;
+
+        // the confirmed state (= what was read from storage most recently)
+        private TState confirmedState;
+        private int confirmedVersion;
+
+        // the handle to the event stream returned by the event store provider
+        private IEventStreamHandle eventStream;
+
+        // must be constructed lazily because grain reference is not available at construction time
+        private IEventStreamHandle EventStream { get {
+                return (eventStream ??
+                   (eventStream = provider.EventStore.GetEventStreamHandle(
+                       provider.EventStore.DefaultStreamName(Host.GetType(), Services.GrainReference))));
+            }
+        }
+
+        public EventStoreLogViewAdaptor(
+            ILogViewAdaptorHost<TState, TEvent> host, 
+            LogConsistencyProvider provider,
+            TState initialState, 
+            ILogConsistencyProtocolServices services)  : base(host, initialState, services)
+        {
+            this.provider = provider;
+
+            if (this.provider == null)
+            {
+                throw new ArgumentException("provider must be an event store", nameof(provider));
+            }
+        }
+
+        // streamname must be constructed lazily because grain reference is not available at construction time
+        private string streamName
+        {
+            get
+            {
+                return EventStream.StreamName;
+            }
+        }
+
+        protected override TState LastConfirmedView()
+        {
+            return confirmedState;
+        }
+
+        protected override int GetConfirmedVersion()
+        {
+            return confirmedVersion;
+        }
+
+        protected override void InitializeConfirmedView(TState initialState)
+        {
+            confirmedState = initialState;
+            confirmedVersion = 0;
+        }
+
+        public override async Task<IReadOnlyList<TEvent>> RetrieveLogSegment(int fromVersion, int toVersion)
+        {
+            var response = await EventStream.Load<TEvent>(fromVersion, toVersion);
+
+            if (response.ToVersion != toVersion || response.FromVersion != fromVersion || response.Events.Count != (toVersion - fromVersion))
+            {
+                throw new NotSupportedException("the event store returned a different segment than was requested");
```

Comment:
InvalidOperationException, as it vialated some invariants, it's not a support thing... or would a different log view adapter support this case?